dynamic range constraint API

avikchaudhuri opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98779

This diff adds the ability to specify range constraints on dynamic dimensions. (Previously we only supported declaring a dynamic dimension, which gets the default range `[2, sympy.oo]`.)

One point worth calling out: our initial design called for compound expressions like `lower <= dynamic_dim(x, d) <= upper`. However this seems difficult to support, because of a combination of desugaring and overloading semantics for such compound expressions in Python. Rather than silently doing the wrong thing, we explicitly error in this case and recommend users to specify multiple constraints, which is supported.

Differential Revision: [D44847318](https://our.internmc.facebook.com/intern/diff/D44847318/)

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

avikchaudhuri committed:
dynamic range constraint API

This diff adds the ability to specify range constraints on dynamic dimensions. (Previously we only supported declaring a dynamic dimension, which gets the default range `[-2, sympy.oo]`.)

One point worth calling out: our initial design called for compound expressions like `lower <= dynamic_dim(x, d) <= upper`. However this seems difficult to support, because of a combination of desugaring and overloading semantics for such compound expressions in Python. Rather than silently doing the wrong thing, we explicitly error in this case and recommend users to specify multiple constraints, which is supported.

Differential Revision: [D44847318](https://our.internmc.facebook.com/intern/diff/D44847318/)

[ghstack-poisoned]

avikchaudhuri committed:
Update on "dynamic range constraint API"


This diff adds the ability to specify range constraints on dynamic dimensions. (Previously we only supported declaring a dynamic dimension, which gets the default range `[-2, sympy.oo]`.)

One point worth calling out: our initial design called for compound expressions like `lower <= dynamic_dim(x, d) <= upper`. However this seems difficult to support, because of a combination of desugaring and overloading semantics for such compound expressions in Python. Rather than silently doing the wrong thing, we explicitly error in this case and recommend users to specify multiple constraints, which is supported.

Differential Revision: [D44847318](https://our.internmc.facebook.com/intern/diff/D44847318/)

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1167,7 +1167,23 @@ def wrap_to_fake_tensor_and_record(
+                            import builtins
+                            dim2constraint[constraint.dim] = StrictMinMaxConstraint(
+                                ValueRanges(
+                                    builtins.max(cr.vr.lower, other_cr.vr.lower),
```
You can't actually assume `other_cr` is a StrictMinMaxConstraint, it might not be...

ezyang commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1167,7 +1167,23 @@ def wrap_to_fake_tensor_and_record(
+                                ValueRanges(
+                                    builtins.max(cr.vr.lower, other_cr.vr.lower),
+                                    builtins.min(cr.vr.upper, other_cr.vr.upper),
+                                )
```
You don't need to do this by hand, ValueRanges supports intersection. `cr.vr & other_cr.vr`

ezyang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -596,9 +602,53 @@ class directly; instead, use :func:`torch._export.dynamic_dim`.
+        # NOTE(avik): We do not support compound expressions like a <= x <= b.
+        # This is because Python implicitly desugars them into bool(a <= x) and bool(x <= b),
+        # and moreover, enforces that any overload of __bool__ must return True or False.
+        # FWIW, sympy also raises TypeError in this case.
```
rip

ezyang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -596,9 +602,53 @@ class directly; instead, use :func:`torch._export.dynamic_dim`.
-    constraint_range: Optional[
-        torch.fx.experimental.symbolic_shapes.StrictMinMaxConstraint
-    ]
+    constraint_range: Optional[StrictMinMaxConstraint]
```
I think you can substantially simplify the logic below if you make this non-Optional

avikchaudhuri replied:
lol you're right, I conservatively, and mistakenly, assumed that `constraint_range=None` actually has a user-facing need, but it doesn't seem so. Killed it, so much simpler.

ezyang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -596,9 +602,53 @@ class directly; instead, use :func:`torch._export.dynamic_dim`.
+    @property
+    def _lower(self):
+        if self.constraint_range is None:
+            return 2
```
e.g., you don't have to do this anymore

avikchaudhuri committed:
Update on "dynamic range constraint API"


This diff adds the ability to specify range constraints on dynamic dimensions. (Previously we only supported declaring a dynamic dimension, which gets the default range `[2, sympy.oo]`.)

One point worth calling out: our initial design called for compound expressions like `lower <= dynamic_dim(x, d) <= upper`. However this seems difficult to support, because of a combination of desugaring and overloading semantics for such compound expressions in Python. Rather than silently doing the wrong thing, we explicitly error in this case and recommend users to specify multiple constraints, which is supported.

Differential Revision: [D44847318](https://our.internmc.facebook.com/intern/diff/D44847318/)

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

avikchaudhuri commented:
@pytorchbot merge -f "bogus lint for line with <100 chars"

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your PR, but lint job is breaking trunk https://hud.pytorch.org/pytorch/pytorch/commit/88dae230d012f2bc834f7bcc7ed9b891ad0763cb and it needs to be fixed before relanding' -c ignoredsignal

malfet commented:
> @pytorchbot merge -f "bogus lint for line with <100 chars"

@avikchaudhuri can you please elaborate what do you mean by that? Pushing a revert directly to trunk.

avikchaudhuri commented:
> > @pytorchbot merge -f "bogus lint for line with <100 chars"
> 
> @avikchaudhuri can you please elaborate what do you mean by that? Pushing a revert directly to trunk.

Sorry. Formatting on VSCode is, I think, configured to 100 chars and the line was well within that, so VSCode wasn't complaining. I didn't realize that we're enforcing 80 chars instead (and this would break trunk), otherwise would have just fixed.
